File: crates/apollo_network/src/bin/broadcast_network_stress_test_node/run/cluster_port_forward.py
#!/usr/bin/env python3
"""
Port forward both Grafana and Prometheus services from Kubernetes cluster to local machine.
"""

import json
import os
import subprocess
import sys
import threading
from cluster_stop import broadcast_network_stress_test_deployment_file_name
from utils import pr, connect_to_cluster
from yaml_maker import prometheus_service_name


def port_forward(service_name, local_port, remote_port, namespace):
    """Run kubectl port-forward for a single service."""
    subprocess.run(
        [
            "kubectl",
            "port-forward",
            f"service/{service_name}",
            f"{local_port}:{remote_port}",
            "-n",
            namespace,
        ]
    )


def main():
    """Set up port forwarding for both Grafana and Prometheus in parallel."""
    if not os.path.exists(broadcast_network_stress_test_deployment_file_name):
        pr("ERROR: No deployment file found. Run cluster_start.py first.")
        sys.exit(1)

    with open(broadcast_network_stress_test_deployment_file_name, "r") as f:
        namespace = json.load(f).get("namespace")

    if not namespace:
        pr("ERROR: No namespace found in deployment file.")
        sys.exit(1)

    connect_to_cluster()

    pr(f"Port forwarding in namespace: {namespace}")
    pr("  → Grafana:    http://localhost:3000")
    pr("  → Prometheus: http://localhost:9090")
    pr("\nPress Ctrl+C to stop\n")

    # Start both port-forwards in separate threads
    threads = [
        threading.Thread(
            target=port_forward,
            args=("grafana-service", 3000, 3000, namespace),
            daemon=True,
        ),
        threading.Thread(
            target=port_forward,
            args=(prometheus_service_name, 9090, 9090, namespace),
            daemon=True,
        ),
    ]

    for t in threads:
        t.start()

    try:
        # Keep main thread alive
        for t in threads:
            t.join()
    except KeyboardInterrupt:
        pr("\nStopping port forwarding...")
        sys.exit(0)


if __name__ == "__main__":
    main()

File: crates/apollo_network_benchmark/src/bin/apollo_network_benchmark_run/cluster_port_forward.rs
use std::process::{Command, Stdio};
use std::thread;

use anyhow::Context;

use crate::mod_utils::{cluster_deployment_file_path, connect_to_cluster, read_deployment_file};
use crate::pr;

const PROMETHEUS_SERVICE_NAME: &str = "prometheus-service";

fn port_forward(service_name: &str, local_port: u16, remote_port: u16, namespace: &str) {
    let _ = Command::new("kubectl")
        .args([
            "port-forward",
            &format!("service/{}", service_name),
            &format!("{}:{}", local_port, remote_port),
            "-n",
            namespace,
        ])
        .stdout(Stdio::inherit())
        .stderr(Stdio::inherit())
        .status();
}

pub fn run() -> anyhow::Result<()> {
    let deployment_data = read_deployment_file(&cluster_deployment_file_path())?;

    let namespace = deployment_data
        .get("namespace")
        .and_then(|n| n.as_str())
        .context("No namespace found in deployment file")?;

    connect_to_cluster()?;

    pr!("Port forwarding in namespace: {}", namespace);
    pr!("  → Grafana:    http://localhost:3000");
    pr!("  → Prometheus: http://localhost:9090");
    pr!("\nPress Ctrl+C to stop\n");

    let namespace_grafana = namespace.to_string();
    let namespace_prometheus = namespace.to_string();

    let grafana_thread = thread::spawn(move || {
        port_forward("grafana-service", 3000, 3000, &namespace_grafana);
    });

    let prometheus_thread = thread::spawn(move || {
        port_forward(PROMETHEUS_SERVICE_NAME, 9090, 9090, &namespace_prometheus);
    });

    let _ = grafana_thread.join();
    let _ = prometheus_thread.join();

    Ok(())
}
